tests/SecretsCrypto/tst_secretscrypto/tst_secretscrypto.c

```diff
@@ -1567,6 +1569,449 @@ static void tst_crypto_import_invalid(SfCryptoFixture *fixture, gconstpointer te
+    g_main_loop_run(fixture->loop);
+
+    encoder = sf_crypto_cipher_session_new_finish(fixture->test_res, &fixture->error);
+
+    g_assert_no_error(fixture->error);
+    g_assert_nonnull(encoder);
+
+    if (g_test_failed())
+        return;
+
+    g_test_queue_unref(encoder);
+
+    sf_crypto_cipher_session_new(fixture->manager,
+            iv,
+            key,
+            SF_CRYPTO_OPERATION_DECRYPT,
+            kd->block_mode,
+            kd->padding,
+            SF_CRYPTO_SIGNATURE_PADDING_NONE,
+            SF_CRYPTO_DIGEST_SHA256,
+            NULL,
+            CRYPTO_PLUGIN_TEST,
+            NULL,
+            _tst_crypto_ref_res_and_quit,
+            fixture);
+    g_main_loop_run(fixture->loop);
+
+    decoder = sf_crypto_cipher_session_new_finish(fixture->test_res, &fixture->error);
+
+    g_assert_no_error(fixture->error);
+    g_assert_nonnull(decoder);
+
+    if (g_test_failed())
+        return;
+
+    g_test_queue_unref(decoder);
+
+    encoded = g_byte_array_sized_new(kd->data_size * kd->chunks);
+    decoded = g_byte_array_sized_new(kd->data_size * kd->chunks);
+
+    g_test_queue_destroy((GDestroyNotify)g_byte_array_unref, encoded);
+    g_test_queue_destroy((GDestroyNotify)g_byte_array_unref, decoded);
+
+    for (i = 0; i < kd->chunks; i++) {
+        GBytes *random_bytes = NULL;
+        GBytes *encoded_bytes;
+
+        sf_crypto_manager_generate_random_data(fixture->manager,
+                kd->data_size,
+                SF_CRYPTO_DEFAULT_CSPRNG_ENGINE,
+                NULL,
+                CRYPTO_PLUGIN_TEST,
+                NULL,
+                _tst_crypto_ref_res_and_quit,
+                fixture);
+        g_main_loop_run(fixture->loop);
+
+        random_bytes = sf_crypto_manager_generate_random_data_finish(fixture->test_res, &fixture->error);
+
+        g_assert_no_error(fixture->error);
+        g_assert_nonnull(random_bytes);
+
+        if (g_test_failed())
+            return;
+
+        g_byte_array_append(encoded, g_bytes_get_data(random_bytes, NULL), g_bytes_get_size(random_bytes));
+
+        sf_crypto_cipher_session_update(encoder,
+                random_bytes,
+                NULL,
+                NULL,
+                _tst_crypto_ref_res_and_quit,
+                fixture);
+        g_bytes_unref(random_bytes);
+        g_main_loop_run(fixture->loop);
+
+        sf_crypto_cipher_session_update_finish(fixture->test_res, &encoded_bytes, &fixture->error);
+
+        g_assert_no_error(fixture->error);
+        g_assert_nonnull(encoded_bytes);
+
+        if (g_test_failed())
+            return;
+
+        sf_crypto_cipher_session_update(decoder,
+                encoded_bytes,
+                NULL,
+                NULL,
+                _tst_crypto_ref_res_and_quit,
+                fixture);
+        g_bytes_unref(encoded_bytes);
+        g_main_loop_run(fixture->loop);
+
+        sf_crypto_cipher_session_update_finish(fixture->test_res, &random_bytes, &fixture->error);
```

Comment:
Are you able to also add a unit test which does multiple updates one-after-another without an update_finish in between each, to ensure that also works?  (I assume that update_finish waits for the dbus response?)

see e.g. https://github.com/sailfishos/sailfish-secrets/blob/master/tests/Crypto/tst_cryptorequests/tst_cryptorequests.cpp#L2103 for example of "batch" encryption mode with cipher session using C++ API.